### DISCUSSION TITLE
NAT Traversal?

### DISCUSSION THREAD

**Discussion by stevefan1999-personal** [siderolabs/talos#8338] [Q&A]
I want to experiment Talos between my home and my Azure network over IPv4, and my home network is obviously having a NAT with a private network range behind a router. More specifically, I want to have my homelab 1 and homelab 2, hidden behind my home network as 192.168.0.1 and 192.168.0.2 respectively, to be able to connect to my Azure machine as 10.10.10.1. So, having a VPN, or rather an overlay network is on the right track and KubeSpan is clearly the choice here.

However, far as I can tell, KubeSpan actually requires public IP addresses for the peers to discover each other (which to be honest, covers 90% of the cases for Kubernetes deployment), and so while it is likely that Wireguard can run as a one-armed bandit if one side has a public IP and the other side doesn't (but that means you would have a unidirectional channel which could slow things down), but if both sides doesn't, no dice.

Adding salt to injury, most home network routers only support symmetric NAT (you can understand it as iptables SNAT/Masquerade if you want to make things easier), so the source port can be rewritten at router level when sending out. Therefore, a proper [NAT traversal](https://en.wikipedia.org/wiki/NAT_traversal) solution is needed for KubeSpan, similar to the kind of [Tailscale](https://github.com/tailscale/tailscale) or [NetBird](https://github.com/netbirdio/netbird), and use [TURN/ICE](https://github.com/coturn/coturn) as a last resort.

**Comment by smira** [comment]
KubeSpan requires one-way connectivity between all machines, so your homelab can be perfectly behind a NAT/firewall, as long as they can connect to each other and Azure workers (over public IP), you're good.

**Reply by stevefan1999-personal** [reply]
Well, if we have a NAT gateway and another instance at Azure, and the virtual network is routed through that NAT gateway? This time, both  Azure instances at KubeSpan would appears to have the same IP, probably the same port as well.

**Reply by stevefan1999-personal** [reply]
As a simple demonstration, maybe we can run Talos at a local computer except on different VM platforms, e.g. run one of them in Hyper-V and one of them in VirtualBox (both using NAT network) to simulate. Scale up if you want to

**Reply by smira** [reply]
What I said above still stays, if there's no way to connect to other machine's IP at least one way, KubeSpan won't help.

E.g. if you deploy Azure VMs with public IP and homelab behind NAT, it's fine.

**Reply by stevefan1999-personal** [reply]
Well, the point is I do have public IPs, except they are shared between different machines behind the scene.

In Azure, I have a NAT gateway with a public IP of 128.0.0.1
And my home router has a public IP of 129.0.0.1. In the home network I have two VMs in 192.168.11.1 and 192.168.11.2. Since most home users would only have one public IP, and would not have a chance to get an assignment. So the KubeSpan endpoint for both home VMs would be 129.0.0.1:51820, right? 

And in Azure I have two Talos VMs in 172.35.0.1 and 172.35.0.2, and they both do not have a public IP on their network interface. As such all the Internet traffic will go through the NAT gateway and masqueraded as 128.0.0.1, thus both VMs would appear to have the same KubeSpan endpoint of 128.0.0.1:51820. How does both Azure VMs and home network VMs reach each other?

So this is why NAT traversal is needed. At the very least we need to try UDP hole punching. Port forwarding would be difficult though.

Of course, if we have IPv6...that's another story.

**Reply by smira** [reply]
Talos needs to know how to connect. Say host A has IPs IP_A_1, IP_A_2, etc. Talos will try all of them with the port 51820. 

If the host has a public IP mapped, Talos will learn it. Talos will not see a hole punched in the NAT. KubeSpan doesn't have a real NAT traversal, nor it has any proxy in the middle atm.

**Reply by stevefan1999-personal** [reply]
Exactly. We are missing the feature to do proper NAT traversal.

**Reply by alongwill** [reply]
@smira is this issue perhaps partly solvable now (since https://github.com/siderolabs/talos/pull/9277) by using https://www.talos.dev/latest/reference/configuration/network/kubespanendpointsconfig/ ?

e.g. if someone has two VMs at home could they configure their router to port forward as follows

```mermaid
graph TD
    Internet((Internet))
    Router[Router<br/>Public IP: 129.0.0.1]
    Talos0[talos-home-0<br/>192.168.11.1:51820]
    Talos1[talos-home-1<br/>192.168.11.2:51820]

    Internet -- "Inbound<br/>52000, 53000" --> Router

    Router -- "Port 52000 →<br/>192.168.11.1:51820" --> Talos0
    Router -- "Port 53000 →<br/>192.168.11.2:51820" --> Talos1

    Talos0 --- Note0[/"KubeSpanEndpointConfig:<br/>129.0.0.1:52000"/]
    Talos1 --- Note1[/"KubeSpanEndpointConfig:<br/>129.0.0.1:53000"/]
```

**Reply by smira** [reply]
yes, as long as you punch the holes in the NAT/firewall, and tell Talos about it.

**Reply by alongwill** [reply]
Thanks. Credit to @TimJones for suggesting it [Slack](https://taloscommunity.slack.com/archives/CMARMBC4E/p1747842901041579?thread_ts=1747840809.216919&cid=CMARMBC4E)

**Comment by steverfrancis** [comment]
One end of the network connection needs to be directly reachable.
So you can have multiple hosts behind your home NAT router - the outgoing
traffic will have distinct source UDP ports, even if all going to 51820 on
the publicly addressed host.
port 51820 does not need to be reachable on all hosts - just in one
direction.

**Comment by stevefan1999-personal** [comment]
I have just tried out my settings, and as I predicted, the connectivity between each KubeSpan node is very fragile at it its best against NAT'd nodes, that most of the time after rebooting the node, you lost the connection to the NAT node almost guaranteed.

**Reply by stevefan1999-personal** [reply]
So yeah I think Talos really need to improve on this, I see that after applying some KubeSpan address filters, the specs and status are not immediately visible
